---
title: "GuideSeq pipeline documentation"
author: "CORRE Guillaume"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---


```{r setup, include=FALSE}
## Global options
knitr::opts_chunk$set(cache = TRUE)
```

# Introduction

# Installation

## Environment

the running environment can be installed using the requirement.yaml file using the command:

`conda create -n guideseq -f requirements.yaml`

Create a folder and clone this git directory

`git clone ....`

Your folder architecture should look similar to :

> | path/to/GuideSeq/
> |     \|-00-pipeline/
> |     \|-01-envs/
> |     \|-02-resources/

# Running the pipeline

## Prepare samples data sheet

The sample data-sheet is a simple delimited file ( ; ) that contains informations about each sample to process in the run. Mandatory columns are:

-   **sampleName** : use to name output files and folders.**Must** be unique across all samples of the run.

-   **Genome**: use to define reference genome. This **must** be one of the values present in config file - **genome** key

-   **gRNA_name** : name the gRNA used.

-   **orientation** : which PCR orientation was chosen ["positive" or "negative"]

-   **gRNA_sequence**: Sequence of the gRNA **without** the PAM sequence

-   **PAM_sequence**: Sequence of the PAM ["NGG"]

-   **Cut_Offset**: Distance from PAM end where the cut occurs [-4]

-   **type**: Type of experiment ["guideseq" or "iguideseq"]. This value will define which sequence to trim.

-   **index1**: Sequence of index 1

-   **index2**: Sequence of index 2

Any column can be added and will be used in the final report.

## Prepare configuration file

The configuration file is a yaml formatted file with key-value dictionary used to fine-tune the pipeline behavior. Settings will apply to all samples of the run.

``` yaml

author: "Guillaume CORRE, PhD"
affiliation: "Therapeutic Gene Editing - GENETHON, INSERM U951, Paris-Saclay University, Evry, France"
version: 0.1


## Library informations
sampleInfo_path: "sampleInfo.csv"

R1: "/media/Data/common/guideseq_gnt_dev/margaux_5/Undetermined_S0_L001_R1_001.fastq.gz"
R2: "/media/Data/common/guideseq_gnt_dev/margaux_5/Undetermined_S0_L001_R2_001.fastq.gz"
I1: "/media/Data/common/guideseq_gnt_dev/margaux_5/Undetermined_S0_L001_I1_001.fastq.gz"
I2: "/media/Data/common/guideseq_gnt_dev/margaux_5/Undetermined_S0_L001_I2_001.fastq.gz"


################################################
## path to references
genome:
  human:
    fasta: "/media/References/Human/Genecode/GRch38/Sequences/GRCh38.primary_assembly.genome.fa"
    index: "/media/References/Human/Genecode/GRch38/Indexes/Bowtie2/GRCh38.primary_assembly.genome" # path to index created during the run if not existing yet
    annotation: "/media/References/Human/Genecode/GRch38/Annotations/gencode.v46.annotation.gtf.gz"
  mouse:
    fasta: ""
    index: ""
    annotation: ""
################################################

################################################
minLength: 25 ## Minimal read length after trimming, before alignment
################################################

## Alignement 
################################################
aligner: "bowtie2"   ## Aligner to use (bowtie2 or bwa)
minFragLength: 100         # Minimal fragment length after alignment
maxFragLength: 1500        # Maximal fragment length after alignment 
################################################

################################################
# post alignment
minMAPQ: 1                      # Min MAPQ score to keep alignments -> !!! multihits have a MAPQ close to 1. Value greater than 1 will discard Offtargets with exact same sequence.
UMI_hamming_distance: 1         # min distance to cluster UMI using network-based deduplication, use [0] to keep raw UMIs
UMI_deduplication: "Adjacency"  # method to correct UMI (cluster or Adjacency)
UMI_pattern: "NNWNNWNN"  
UMI_filter: "FALSE"               # If TRUE, remove UMIs that do no match the expected pattern [FALSE or TRUE]
################################################


################################################
## Off targets calling
tolerate_bulges: "FALSE"           # whether to include gaps in the gRNA alignment (this will change the gap penalty during SW pairwise alignment)
max_edits_crRNA: 6              # filter clusters with less or equal than n edits in the crRNA sequence (edits = substitutions + INDELs)
ISbinWindow: 100                # insertion sites closer than 'ISbinWindow' will be clustered together
minReadsPerUMI: 0               # 0 to keep all UMIs, otherwise min number of reads per UMIs
minUMIPerIS: 0                  # 0 to keep all IS, otherwise min number of UMI per IS
slopSize: 50                    # window size (bp) around IS (both directions) to identify gRNA sequence (ie 50bp = -50bp to +50bp)
min_predicted_distance: 100     # distance between cut site and predicted cut site to consider as predicted
################################################

################################################
# reporting
max_clusters: 100                 # max number of cluster alignments to report
minUMI_alignments_figure: 1       # filter clusters with more than n UMI in the report alignment figure (set to 0 to keep all clusters -> can be slow)
################################################

# Prediction
################################################
SWoffFinder:
  path: "/opt/SWOffinder" ## Path to SWoffinder on your server (downloaded from https://github.com/OrensteinLab/SWOffinder)
  maxE: 6                 # Max edits allowed (integer).
  maxM: 6                 # Max mismatches allowed without bulges (integer).
  maxMB: 6                # Max mismatches allowed with bulges (integer).
  maxB: 3                 # Max bulges allowed (integer).
  window_size: 100
################################################



################################################
# Sequences for the trimming steps

guideseq:
  positive:
    R1_trailing: "GTTTAATTGAGTTGTCATATGT"
    R2_leading: "ACATATGACAACTCAATTAAAC"
    R2_trailing: "AGATCGGAAGAGCGTCGTGT"
  negative:
    R1_trailing: "ATACCGTTATTAACATATGACAACTCAA"
    R2_leading: "TTGAGTTGTCATATGTTAATAACGGTAT"
    R2_trailing: "AGATCGGAAGAGCGTCGTGT"



iguideseq:
  positive:
    R2_leading: "ACATATGACAACTCAATTAAACGCGAGC"
    R2_trailing: "AGATCGGAAGAGCGTCGTGT"
    R1_trailing: "GCTCGCGTTTAATTGAGTTGTCATATGT"
  negative:
    R1_trailing: "TCGCGTATACCGTTATTAACATATGACAACTCAA"
    R2_leading: "TTGAGTTGTCATATGTTAATAACGGTATACGCGA"
    R2_trailing: "AGATCGGAAGAGCGTCGTGT"
################################################
```

## Folder structure

In order to start a run:

-   create a new directory in the installation folder and `cd` into.

-   Then :

    -   move the sample data-sheet

    -   move the configuration file

    -   move the illumina sequencing `undeterminded_R1/R2/I1/I2.fastq.gz` files (undemultiplexed)

> Input fastq files should respect the following structure from original paper :
>
> -   R1 : contains fragment sequence starting in gDNA
>
> -   R2: Starts with ODN sequence followed by gDNA sequence and potential adaptor sequence
>
> -   i1 : Contains barcode 1 (usually 8 nucleotides)
>
> -   i2 : Contains barcode 2 and UMI sequence (usually 8 + 8 nulceotides)

Your folder architecture should look similar to :

> | path/to/GuideSeq/
> |     \|-00-pipeline/
> |     \|-01-envs/
> |     \|-02-resources/
> |     **\|-my_analysis_folder**
> |         \|-sampleInfo.csv
> |         \|-guideSeq_GNT.yml
> |         \|- undertermined_R1.fastq.gz
> |         \|- undertermined_R2.fastq.gz
> |         \|- undertermined_I1.fastq.gz
> |         \|- undertermined_I2.fastq.gz

From inside your analysis folder, run the command below after adjusting for number of CPU (-j) and

```         
snakemake -s ../00-pipeline/guideseq_gnt.smk \
  -j 24 \
  -k \
  --use-conda \
  --conda-prefix path/to/conda/envs \
  --conda-frontend mamba \
  -n
```

# Output

## Report

## Tables

# Pipeline step-by-step explanations

## Demultiplexing

**Tool** : cutadapt

Undetermined fastq files are demultiplexed to `sampleName` fastq files according to barcodes present in the sample data sheet.

-   Barcode1 and barcode2 sequences are concatenated to build the demultiplexing index.

-   i1 and i2 fastq files are concatenated to a single fastq file (i3)

-   R1 and R2 fastq are demultiplexed according to i1+i2 sequence present in i3 fastq files.

-   UMI sequence is added to R1 and R2 read name for future UMI quantification.

    -   UMI is extracted from i3 first nucleotides according to the length of `UMI_pattern` variable in the configuration file.

## ODN trimming:

**Tool** : cutadapt

The leading ODN sequence is remove from R2 reads according to the `method` and `PCR orientation` defined in the sample data sheet for each sample.

Reads that **do not start** with the ODN sequence are discarded.

## Adaptor trimming:

**Tool** : cutadapt

The ODN and adaptor trailing sequences are trimmed from R1 and R2 reads respectively if present. If the DNA fragment is large compared to R1/R2 sequences length, those sequences may not be present.

## Read filtering:

**Tool** : cutadapt

After trimming, only read pairs with **both** mates longer than a `minLength` defined in the configuration file are selected for alignment.

## Genome alignment:

Reference sequences are specified in the configuration file. Genome index is build if it does not already exist.

``` yaml
## path to references
genome:
  human:
    fasta: "/media/References/Human/Genecode/GRch38/Sequences/GRCh38.primary_assembly.genome.fa"
    index: "/media/References/Human/Genecode/GRch38/Indexes/Bowtie2/GRCh38.primary_assembly.genome" # path to index created during the run if not existing yet
    annotation: "/media/References/Human/Genecode/GRch38/Annotations/gencode.v46.annotation.gtf.gz"
```

Trimmed reads pairs that passed the filtering steps are then aligned on the `reference genome` specified in the sample data sheet for each sample using the `aligner` specified in configuration file.

``` yaml
## Alignement 
################################################
aligner: "bowtie2"   ## Aligner to use (bowtie2 or bwa)
minFragLength: 100         # Minimal fragment length after alignment
maxFragLength: 1500        # Maximal fragment length after alignment 
################################################
```

Multihit alignments with low MAPQ score are discarded except if the Alignment score (AS tag) is equal to the score of the second best alignment score (XS tag) .

### Multi-hits management:

Multihits are reads with multiple possible equally good alignment positions in the genome. We choose to keep only a single random alignment for each read instead of reporting all possible positions.

On the long run, if multi read arise from the same cuting site, they will distribute randomly to all sites (explain more).

## Cutting site calling:

Following alignment on the reference genome, nuclease cutting sites are extracted from the start position of R2 read alignment.

Reads that align at the same cutting site with the same UMI are aggregated together keeping track of total reads count.

## UMI correction:

UMI sequences are corrected for potential sequencing error using the parameters defined in the configuration file

``` yaml
UMI_hamming_distance: 1         # min distance to cluster UMI using network-based deduplication, use [0] to keep raw UMIs
UMI_deduplication: "Adjacency"  # method to correct UMI (cluster or Adjacency)
UMI_pattern: "NNWNNWNN"  
UMI_filter: "FALSE"               # If TRUE, remove UMIs that do no match the expected pattern [FALSE or TRUE]
```

## Cut site clustering:

Cut sites than fall in the same window of `ISbinWindow` defined in the configuration file are clustered together.

## gRNA match:

For each cluster of cutting sites, the gRNA sequence defined in the sample data sheet for each sample is looked up in a window of +/- `slopSize` bp using the Swith-Watterman algorithm.

Gap tolerance can be accepted to detect bulges in gDNA or gRNA if the `tolerate_bulges` variable is set.

## Annotation of clusters:

Clusters of cutting sites are annotated using the gtf file specified in the configuration file for each organism.

A first step prepare the annotation file to extract only gene and exons features.

A second step annotate clusters to gene and position relative to those genes (exon/intron). Multiple annotations may be present for each cluster and are reported.

## Off target prediction:

for each gRNA sequence and each organism specified in the sample data sheet, a prediction of OTS is realized using the SWOffinder tool with parameters defined in the configuration file :

``` yaml
# Prediction
################################################
SWoffFinder:
  path: "/opt/SWOffinder" ## Path to SWoffinder on your server (downloaded from https://github.com/OrensteinLab/SWOffinder)
  maxE: 6                 # Max edits allowed (integer).
  maxM: 6                 # Max mismatches allowed without bulges (integer).
  maxMB: 6                # Max mismatches allowed with bulges (integer).
  maxB: 3                 # Max bulges allowed (integer).
  window_size: 100
################################################
```

## Reporting:

A short report is generated with main tables and graphical representations to better understand pipeline results.
