---
title: "GuideSeq report"
author: "`r config$author`"
date: "`r Sys.Date()`"
subtitle: "`r dirname(config_path)`"
output: 
    rmdformats::readthedown:
    highlight: kate
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,cache = F)
options(verbose = F)
options(knitr.kable.NA = '')
```

```{r,error=F,message=FALSE,warning=FALSE,echo=FALSE}
# Libraries

library(tidyverse,quietly = T,warn.conflicts = F,verbose = F)
library(kableExtra,quietly = T,warn.conflicts = F,verbose = F)
library(yaml,quietly = T,warn.conflicts = F,verbose = F)

libraries_count <- length(tables_html)

```

# Description of libraries

```{r,echo=F}
kbl(sampleInfo %>% 
      arrange(sampleName) %>% 
      select(library=sampleName,
             replicate,
             "Cells"=Cell_Type, 
             Genome,"gRNA"=gRNA_name,
             "gRNA sequence"=gRNA_sequence, 
             "PAM"=PAM_sequence,
             Cas,
             Cut_Offset ,
             type,
             orientation) %>% 
      select(
    where(
      ~sum(!is.na(.x)) > 0
    )),
                      caption = "Table 1") %>%
        kable_classic_2(full_width = T,html_font = "helvetica") %>% 
        kable_styling(bootstrap_options = c("condensed","hover","stripped"),
                      font_size = 11,
                      fixed_thead = T) 
```

# Best match found

```{r,echo=F}
kbl(best_aligns %>% arrange(library) %>% select(-clusterID,-cut_modal_position) %>% unite("position",chromosome,cut_gRNa_alignment,sep=":"),
                      caption = "Table 1b",
    format.args = list(big.mark = ",")) %>%
        kable_classic_2(full_width = T,html_font = "helvetica") %>% 
        kable_styling(bootstrap_options = c("condensed","hover","stripped"),
                      font_size = 12,
                      fixed_thead = T) %>%
  collapse_rows(columns = 1, valign = "middle")
```

# Statistics of reads processing

Reads were processes according to the following steps :

-   Reads were demultiplexed using **cutadapt**.

-   Presence of the ODN sequence at the 5' end of the R2 reads was checked.

-   I5 and I7 adaptors were trimmed if present

-   Reads longer than **`r config$minLength`** bp were filtered (both of the pair).

Table 2 presents the read counts and % of demultiplexed for each step of the read processing.

```{r, echo = F}
kbl(stats%>% arrange(library),
                      caption = "Table 2") %>%
        kable_classic_2(full_width = T,html_font = "helvetica") %>% 
        kable_styling(bootstrap_options = c("condensed","hover","stripped"),
                      font_size = 12,
                      fixed_thead = T)
```

# Statistics of read alignment

## Genome mapping

Reads passing the QC were aligned on the reference genome **`r basename(config$genome$human$fasta)`** using **`r config$aligner`** (see [Configuration settings]) .

## Cutting sites calling and clustering

Cut sites were identified from the alignment start position of R2 reads.

-   Reads aligning to the exact same start position were aggregated if they share the same UMI sequence.

-   UMI were corrected using the **`r config$UMI_deduplication`** method with a Hamming distance tolerance of **`r config$UMI_hamming_distance`**.

-   Positions/UMI combinations with more than **`r config$minReadsPerUMI`** reads were considered for next step.

Clusters are defined as a group of ODN insertion sites within a distance smaller than **`r config$ISbinWindow`** bp.

Clusters are then characterized by :

-   Whether there are bi-directional reads alignments indicating both ODN orientations ("dual Orientation" in table below)

-   The number of ODN unique insertion sites ("multiple Cuts" in table below)

-   The presence of the crRNA sequence with less than **`r config$max_edits_crRNA`** edits withing the cluster boundaries +/- **`r config$slopSize`** bp("crRNA matched" in table below)

Clusters with more than **`r config$minUMIPerIS`**total UMIs were considered.

```{r,echo=F,warning=FALSE, message=FALSE,error=FALSE}
alignment_table <- lapply(tables_off,data.frame) %>%
  bind_rows(.id="library")

names(alignment_table) <- c('library','link')

alignment_table <- alignment_table %>% 
  arrange(library) %>%  
  mutate(Alignments = cell_spec("link", "html", new_tab=T,link = paste("./report-files/",basename(link),sep = ""))) %>% 
  select(-link)



kbl(stats_summary %>%  arrange(library) %>% left_join(alignment_table),escape = F,
                      caption = "Table 3",
    format.args = list(big.mark = ",")) %>%
        kable_classic_2(full_width = T,html_font = "helvetica") %>% 
        add_header_above(c(" " = 4, "Clusters" = 5)) %>% 
        kable_styling(bootstrap_options = c("condensed","hover","stripped"),
                      font_size = 12,
                      fixed_thead = T)
```

# Graphical representations

## Number of clusters per chromosome

This figure represents the distribution of unique clusters per chromosome and library, colored by the presence of the gRNA

```{r,fig.width=12, fig.height=ceiling(libraries_count/3)*4,fig.align='center',fig.format='png', echo=FALSE}
fig3
```

This figure represents the distribution of unique clusters per chromosome and library, with gRNA match and colored by its prediction status.

```{r,fig.width=12, fig.height=ceiling(libraries_count/3)*4,fig.align='center',fig.format='png', echo=FALSE}
fig4
```

## Number of UMI per chromosome

This figure represents the total number of UMIs (cells) per chromosome and library, colored by the presence of the gRNA

```{r,fig.width=12, fig.height=ceiling(libraries_count/3)*4,fig.align='center',fig.format='png', echo=FALSE}
fig3b
```

This figure represents the total number of UMIs (cells) per chromosome and library, for clusters with gRNA match and colored by prediction status.

```{r,fig.width=12, fig.height=ceiling(libraries_count/3)*4,fig.align='center',fig.format='png', echo=FALSE}
fig4b
```

## Rank-abundance curve

The rank-abundance curve is used in ecology to visualize species diversity in a community by plotting species abundance against species rank on logarithmic scales. The resulting curve's shape provides insights into relative species abundance patterns and community evenness - a steep curve indicates dominance by a few species, while a shallow curve suggests more even distribution of abundances across species.

Only clusters with gRNA match and **`r config$minUMIPerIS`** total UMIs are plotted. Red dots correspond to clusters with 0 Edits in the crRNA and PAM sequences. Top3 most abundant clusters (UMI counts) are labeled.

```{r,fig.width=12, fig.height=ceiling(libraries_count/3)*3,fig.align='center',fig.format='png', echo=FALSE}
fig5
```

## Distribution of cut sites around on-Target sites

```{r,fig.width=12,fig.height=ceiling(nrow(best_aligns)/3)*2,fig.align='center',fig.format='png', echo=FALSE}
fig7
```

## Distribution of the number of edits in the crRNA

The figure represents the distribution of the number of edits in the crRNA sequence per cluster with gRNA match, colored position prediction status.

```{r,fig.width=12, fig.height=ceiling(libraries_count/3)*2,fig.align='center',fig.format='png', echo=FALSE}
fig6
```

```{r,fig.width=12, fig.height=ceiling(libraries_count/3)*2,fig.align='center',fig.format='png', echo=FALSE,eval=FALSE}
fig6b
```

# Configuration settings

``` yaml
`r xfun::file_string(config_path)`
```

# R session informations

```{r}
sessionInfo()
```
